text/0000-secret-types.md

```diff
@@ -0,0 +1,206 @@
+
+[1] *Cryptographic Software in a Post-Spectre World.* Chandler Carruth. RWC 2020. Recording forthcoming.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Secret integer types are a type of restricted integer primitive. In particular, non-constant time operations like division are prohibited. Printing of secret integers directly is also prohibited--they must first be declassified into non-secret integers.
+
+These integers are intended to form a basis for the creation of other secret types, which can be built on top of them, such as secret strings for storing passwords in them.
+
+Comparison of secret integer types is allowed, but the algorithm is independent of the secrets. Comparison of secret types must return a secret_bool, which cannot be branched on.
+
+Example: comparison
+```
+let x : secret_i8 = 6;
+Let y : secret_i8 = 10;
+if (x < y) { // compiler error: cannot branch on secret bool
+  ...
+}
+```
+Example: declassification
+```
+let x : secret_i8 = 6;
+let y : secret_i8 = 10;
+println!((x ^ y).declassify())
+```
+
+Since indexing vectors and arrays is only possible using `usize,` you will be prevented from indexing into a vector or array with a secret integer. Error messages will derive naturally from the type system.
+
+
+Secret types will likely be a more advanced topic when teaching Rust.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+For each fixed-size integer type:
+Implement the following methods:
+- From_be
+- From_le
+- From_be_bytes
+- From_le_bytes
+- From_ne_bytes
+- Is_positive
+- Is_negative
+- Leading_zeros
+- Min_value
+- Max_value
+- Overflowing_add
+- Overflowing_sub
+- Overflowing_mul
+- Overflowing_neg
+- Overflowing_shl
+- Overflowiing_shr
+- Overflowing_pow
+- Pow
+- Reverse_bits
+- Rotate_left
+- Rotate_right
+- Saturating_add
+- Saturating_sub
+- Saturating_neg
+- Saturating_mul
+- Saturating_pow
+- Signum
+- Swap_bytes
+- To_be
+- To_le
+- To_be_bytes
+- To_le_bytes
+- To_ne_bytes
+- Trailing_zeros
+- Wrapping_add
+- Wrapping_sub
+- Wrapping_mul
+- Wrapping_neg
+- Wrapping_shl
+- Wrapping_shr
+- Wrapping_pow
+
+
+Implement the following traits
+Secret integers may only  be combined with other secret integers and the result will be a secret type
+- Add
+- AddAssign
+- BitAnd
+- BitAndAssign
+- BitOr
+- BitOrAssign
+- BitXor
+- BitXorAssign
+- Clone
+- Copy
+- Default
+- Drop
+- Mul
+- MulAssign
+- Neg
+- Not
+- Shl
+- ShlAssign
+- Shr
+- ShrAssign
+- Sub
+- SubAssign
+- Sub
+
+
+For secret_bool, implement the following methods:
+- n/a
+And the following traits
+- BitAnd
+- BitAndAssign
+- BitOr
+- BitOrAssign
+- BitXor
+- BitXorAssign
+- Clone
+- Copy
+- Default
+- Not
+
+We will also need to define a trait Classify<T> for T and method declassify. Classify will take a non secret integer or boolean and return a secret integer or boolean. `declassify` will consume a secret integer returning a non-secret integer.
```

Comment:
I think it would make for a more readable proposal if you specified the trait definitions, and the other operations in ```rust blocks.